---
output:
  md_document:
    variant: markdown_github
---

<!-- badges: start -->
[![R-CMD-check](https://github.com/fmicompbio/TabulaMurisSenisData/workflows/R-CMD-check/badge.svg)](https://github.com/fmicompbio/TabulaMurisSenisData/actions)
<!-- badges: end -->

# TabulaMurisSenisData 

This is an R/[ExperimentHub](https://bioconductor.org/packages/release/bioc/html/ExperimentHub.html) package providing easy access to bulk and single-cell RNA-seq data from the [Tabula Muris Senis](https://tabula-muris-senis.ds.czbiohub.org/) project. 
The package is under submission to Bioconductor. 
For now, it can be installed from this GitHub repository using the [devtools](https://cran.r-project.org/package=devtools) package:

```{r, eval=FALSE}
devtools::install_github("fmicompbio/TabulaMurisSenisData")
```

Note that R version 4.1 or newer, and Bioconductor release 3.14 (current devel version) or newer are required to use the package.

```{r}
library(TabulaMurisSenisData)
```

# Available data

The package provides access to the [bulk RNA-seq data](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132040) as well as the [droplet](https://figshare.com/articles/dataset/Processed_files_to_use_with_scanpy_/8273102?file=23938934) and [FACS](https://figshare.com/articles/dataset/Processed_files_to_use_with_scanpy_/8273102?file=23937842) single-cell RNA-seq data sets from the Tabula Muris Senis project. 
Each data set is returned as a SingleCellExperiment object. 
For the single-cell data sets, we also provide separate objects for the individual tissues ([droplet](https://figshare.com/articles/dataset/Tabula_Muris_Senis_Data_Objects/12654728) and [FACS](https://figshare.com/articles/dataset/Tabula_Muris_Senis_Data_Objects/12654728), respectively).
No additional analysis has been performed on the data.

## Bulk RNA-seq

The bulk RNA-seq data is accessed via the `TabulaMurisSenisBulk()` function. 

```{r}
(tms_bulk <- TabulaMurisSenisBulk())

## List available tissues
table(tms_bulk$organ)
```

## Droplet scRNA-seq

The droplet scRNA-seq data is accessed via the `TabulaMurisSenisDroplet()` function. 
By default, the full data set will be returned.
Data from each individual tissue can also be obtained by specifying the `tissues` argument appropriately. 
Note that the return value from this function is always a named list, with one element per retrieved tissue (or 'All' for the full data set).

```{r}
(tms_droplet <- TabulaMurisSenisDroplet(tissues = "All"))

listTabulaMurisSenisTissues(dataset = "Droplet")
```

The returned objects contain also the reduced dimension representations generated by the Tabula Muris Senis consortium (see the vignette for more details on how to recreate the plot below from the provided data).

![](inst/images/droplet-umap.png)

## FACS scRNA-seq

The FACS scRNA-seq data is accessed via the `TabulaMurisSenisFACS()` function. 
As for the droplet dat, the full data set will be returned by default, but individual tissue data sets are also available.

```{r}
(tms_facs <- TabulaMurisSenisFACS(tissues = "All"))

listTabulaMurisSenisTissues(dataset = "FACS")
```

![](inst/images/facs-umap.png)

# Rendered processing reports

The following reports provide detailed descriptions of how the data from the Tabula Muris Consortium was obtained and processed for inclusion in this package. 

* [bulk data](https://www.dropbox.com/s/4c8jplihh1nyj4u/make-data-tabula-muris-senis-bulk.html?dl=0)
* [droplet single-cell data](https://www.dropbox.com/s/xckngw1u4aqi3xa/make-data-tabula-muris-senis-droplet.html?dl=0)
* [FACS single-cell data](https://www.dropbox.com/s/gs47snnkstinm2j/make-data-tabula-muris-senis-facs.html?dl=0)
